core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -881,11 +905,20 @@ private[spark] object MapOutputTracker extends Logging {
         logError(errorMessage)
         throw new MetadataFetchFailedException(shuffleId, startPartition, errorMessage)
       } else {
-        for (part <- startPartition until endPartition) {
-          val size = status.getSizeForBlock(part)
-          if (size != 0) {
+        if (endPartition - startPartition > 1 && supportsContinuousBlockBatchFetch) {
```

Comment:
nit: we can define a `val numBlocks = endPartition - startPartition`, as it's used more than once